### PR TITLE
Extending RNG API started in symplectic manifold

### PR DESCRIPTION
**Pull request by mateuszbaran**
The symplectic manifold PR: #447 extended the `rand` function to generate random points and tangent vectors on a manifold. This is useful and I think we can extend it to other manifolds.

The limitation here is that I don't want to make any guarantees regarding the distribution that is sampled. @kellertuer if you're OK with this I'll add here more methods, possibly even making it a replacement for the random point generation in Manopt.

### CODE REVIEW

**Comment by mateuszbaran** [comment]
Sure, I'll add tangent vector generation as well.

**Comment by kellertuer** [comment]
I like the `vector_at` idea. My first idea would have been `rand(M,p)` but that breaks with the usual interface of rand only having one argument, right? Alternatively one could also do `rand(TangentSpaceAt(M,p))` or `rand((M,p))`. What do you think?

**Comment by mateuszbaran** [comment]
> I like the `vector_at` idea. My first idea would have been `rand(M,p)` but that breaks with the usual interface of rand only having one argument, right? Alternatively one could also do `rand(TangentSpaceAt(M,p))` or `rand((M,p))`. What do you think?

Actually the standard meaning of `rand(M, p)` is "give me p points from M". I'll add support for `rand(TangentSpaceAt(M,p))` anyway, and `rand((M,p))` looks a bit weird to me. `vector_at` is what we've agreed upon in the symplectic PR.

**Comment by kellertuer** [comment]
Ok, then we use `vector_at`.

**Comment by mateuszbaran** [comment]
I think this now covers all features from Manopt.jl random point/tangent vector generation, and a bit more (in-place variants, RNG state passing and points on Hyperbolic).

File: src/manifolds/SymmetricPositiveDefinite.jl [hunk 1]
```diff
@@ -156,6 +156,80 @@ project(::SymmetricPositiveDefinite, p, X)
...
+Generate a random symmetric positive definite matrix on the
+`SymmetricPositiveDefinite` manifold `M`.
+"""
+rand(M::SymmetricPositiveDefinite; σ::Real=1)
+
+function Random.rand!(
+    M::SymmetricPositiveDefinite{N},
+    pX;
+    vector_at=nothing,
+    σ::Real=one(eltype(pX)) / (vector_at === nothing ? 1 : 100),
```
[the remark this answers is not in the document]
Comment (mateuszbaran) [reply src/manifolds/SymmetricPositiveDefinite.jl:171]: I don't really know, I copied it from Manopt: https://github.com/JuliaManifolds/Manopt.jl/blob/master/src/helpers/random.jl#L355 .

[the remark this answers is not in the document]
Comment (kellertuer) [reply src/manifolds/SymmetricPositiveDefinite.jl:171]: Ah, that guy - he always writes strange stuff ;)
One reason might be that the “randomness” had a too large variance usually, at least for Rician noise (which you linked) - ah I see for Gaussian as well. What we could do by default in the linear affine norm here is something inspired by the 1D case where the norm is XY/p^2, so we could set that to `1/norm(p,'fro')` ?

[the remark this answers is not in the document]
Comment (mateuszbaran) [reply src/manifolds/SymmetricPositiveDefinite.jl:171]: Sure, I can divide it by `norm(vector_at)` instead (which is the Frobenius norm).

[the remark this answers is not in the document]
Comment (kellertuer) [reply src/manifolds/SymmetricPositiveDefinite.jl:171]: ...no not of the vector, of p (or whoever its called `pX` here).

[the remark this answers is not in the document]
Comment (mateuszbaran) [reply src/manifolds/SymmetricPositiveDefinite.jl:171]: No, `pX` is where the output is written (this is a mutating function). We are supposed to return a vector from the tangent space at point `vector_at`.

[the remark this answers is not in the document]
Comment (kellertuer) [reply src/manifolds/SymmetricPositiveDefinite.jl:171]: oh, sorry I confused that and thought `vector_at` would be a vector, sorry (long day), you are of course right.

[the remark this answers is not in the document]
Comment (mateuszbaran) [reply src/manifolds/SymmetricPositiveDefinite.jl:171]: No problem, `vector_at` is not a particularly intuitive name.

[the remark this answers is not in the document]
Comment (kellertuer) [reply src/manifolds/SymmetricPositiveDefinite.jl:171]: maybe `at_point` or `tangentspace_at` would be better?

[the remark this answers is not in the document]
Comment (mateuszbaran) [reply src/manifolds/SymmetricPositiveDefinite.jl:171]: Hm, I don't know. None of them seem particularly convincing but maybe I'll leave `vector_at` for now.

[the remark this answers is not in the document]
Comment (mateuszbaran) [reply src/manifolds/SymmetricPositiveDefinite.jl:171]: After all, `vector_at` is already in use.

[the remark this answers is not in the document]
Comment (kellertuer) [reply src/manifolds/SymmetricPositiveDefinite.jl]: Ok, that's a good reason to keep it.

**Comment by mateuszbaran** [comment]
Can we merge this before #468 ? These PRs should be mostly independent anyway.

**Review by kellertuer — APPROVED** [review]
To me this looks fine.

**Pull request merged**